### ISSUE TITLE
Remove ` no-array-for-each` from `unopinionated` preset

### ISSUE THREAD

**Issue by sickhippie** [sindresorhus/eslint-plugin-unicorn#2820]
### Description

SonarQube/SonarCloud just added 68 rules from this plugin to their SonarWay ruleset, which is how I ended up here.  While my main complaint is over there, my research led me down this rabbit hole and to @fisker suggesting to open an issue for rules that someone thinks are actually opinionated to have a discussion about it.

Looking at the original ticket and PR explanation, rules that were specifically stylistic or opinionated changes should not have been included.  `no-array-reduce` and `no-for-loop` were not added as they are stylistic choices.

When the rule `no-array-for-each` [was added](https://github.com/sindresorhus/eslint-plugin-unicorn/issues/955), the main justification was stylistic: "`for...of` is more readable."  Readability is almost always a personal preference.  I find `forEach` significantly more readable than `for...of`, especially if the index is needed in the loop.  "Github disallows this too" was added as a supporting argument, but [this comment](https://github.com/github/eslint-plugin-github/issues/86#issuecomment-579677436) suggests it's inclusion there is for more org-specific reasons than any larger purpose.

`for...of` is not always faster than `forEach`, and can perform much worse as arrays get larger until you're in the millions of items range.  With small arrays, the difference in performance is (usually) tiny.  Being able to use control flow to break/continue is not a concern for most uses of `forEach`, and if those are necessary for the task then `for...of` is still an available tool.  That becomes a "right tool for the job" decision best left up to the developer.

I have yet to see a scenario where typescript went weird inside `forEach` that didn't uncover a logical or programmatical issue that was at fault.  I'm not saying that hasn't happened, but if it's not popped up my decade of heavy TS work it doesn't feel enough of a concern to complete disallow `forEach`.

The one scenario that can bubble up is that `forEach` isn't async-safe (like most Array iteration methods), but those scenarios are handled in a larger ruleset like `@typescript-eslint/no-misused-promises`.

### Examples

```js
Difficult to read & parse
// ❌
for (const [index, element] of list.entries()) {
  console.log(`${element} of ${index}`);
}

Easy to read & parse
// ✅
list.forEach((element, index) => {
  console.log(`${element} of ${index});
}
```

**Comment by sindresorhus** [comment]
After some considering, we’re going to keep `no-array-for-each` in `unopinionated`.

The main reason is that this rule is not just about whether someone finds `forEach` readable. `forEach` moves iteration into a callback, which changes the shape of the code in ways that matter: no `break` or `continue`, awkward `await` behavior, callback scoping, and weaker TypeScript control-flow narrowing. The narrowing example above is a good illustration. With `for...of`, the compiler and the reader stay in the same control-flow model as the surrounding code.

That makes this different from `no-for-loop` and `no-array-reduce`. Those rules mostly push a project toward a particular style of expressing iteration or accumulation. `no-array-for-each` pushes away from an API that looks like ordinary iteration but lacks ordinary loop semantics. That is a real footgun, not just a formatting preference.

`unopinionated` also does not mean “no judgement calls”. It means we reserve the preset for rules we think are broadly useful, have low false-positive cost, and improve code in ways that are not primarily cosmetic. This rule still fits that bar. Projects that intentionally prefer callback iteration can disable this one rule, but we don’t think that project preference should remove the rule from the shared preset.
